src/animation.rs

```diff
@@ -0,0 +1,122 @@
+use keyframe::{ease_with_scaled_time, functions::EaseInOut};
+use std::{
+    collections::HashMap,
+    time::{Duration, Instant},
+};
+
+#[derive(Default)]
+pub struct Animator {
+    animations: HashMap<Animation, AnimationInner>,
+    remove_later: Vec<Animation>,
+}
+
+#[derive(Debug)]
+struct AnimationInner {
+    start: f64,
+    end: f64,
+    current: f64,
+
+    start_time: Instant,
+    duration: Duration,
+}
+
+#[derive(Eq, PartialEq, Debug, Hash, Clone)]
+pub enum Animation {
+    Height,
+}
+
+pub struct AnimationConfig {
+    pub height: Duration,
+}
+
+impl Animator {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_animation(&mut self, name: Animation, start: f64, end: f64, duration: Duration) {
+        match self.animations.entry(name) {
+            std::collections::hash_map::Entry::Occupied(o) =>
+                panic!("Add already existed animation with {:?}. It is a bug. Please submit a bug to https://github.com/l4l/yofi/issues", o),
+            std::collections::hash_map::Entry::Vacant(v) => {
+                v.insert(AnimationInner::new(start, end, duration));
+            }
+        }
+    }
+
+    pub fn cancel_animation(&mut self, name: Animation) {
+        self.animations.remove(&name);
+    }
+
+    pub fn contains(&self, name: Animation) -> bool {
+        self.animations.contains_key(&name)
+    }
+
+    pub fn get_value(&mut self, name: Animation) -> Option<f64> {
+        Some(self.animations.get(&name)?.current)
+    }
+
+    pub fn proceed(&mut self) -> bool {
+        let current_time = Instant::now();
+
+        for name in std::mem::take(&mut self.remove_later).into_iter() {
+            self.animations.remove(&name).unwrap();
```

Comment:
You could get a panic here, reproduction:

`seq 100 | cargo r dialog` (debug build is important)

then quickly type any two digits